tuta-sdk/rust/src/json_serializer.rs

```diff
@@ -0,0 +1,483 @@
+        if model_value.cardinality == Cardinality::ZeroOrOne && json_value == JsonElement::Null {
+            return Ok(ElementValue::Null);
+        }
+
+        // Type models for ids are special.
+        // The actual type depends on the type of the Element.
+        // e.g. for ListElementType the GeneratedId actually means IdTuple.-
+        if value_name == "_id" {
+            return match (
+                &model_value.value_type,
+                json_value,
+                &type_model.element_type,
+            ) {
+                (
+                    ValueType::GeneratedId | ValueType::CustomId,
+                    JsonElement::String(v),
+                    ElementType::Element | ElementType::Aggregated,
+                ) => Ok(ElementValue::String(v)),
+                (
+                    ValueType::GeneratedId | ValueType::CustomId,
+                    JsonElement::Array(arr),
+                    ElementType::ListElement,
+                ) if arr.len() == 2 => match Self::parse_id_tuple(arr) {
+                    None => invalid_value(),
+                    Some(id_tuple) => Ok(ElementValue::IdTupleId(id_tuple)),
+                },
+                _ => invalid_value(),
+            };
+        }
+
+        match (&model_value.value_type, json_value) {
+            (ValueType::String, JsonElement::String(v)) => Ok(ElementValue::String(v)),
+            (ValueType::Number, JsonElement::String(v)) => match v.parse::<i64>() {
+                Ok(num) => Ok(ElementValue::Number(num)),
+                Err(_) => invalid_value(),
+            },
+            (ValueType::Bytes, JsonElement::String(v)) => {
+                let vec = match BASE64_STANDARD.decode(v) {
+                    Ok(v) => Ok(v),
+                    Err(_) => Err(InvalidValue {
+                        type_ref: type_model.into(),
+                        field: value_name.to_owned(),
+                    }),
+                }?;
+                Ok(ElementValue::Bytes(vec))
+            }
+            (ValueType::Date, JsonElement::String(v)) => {
+                let num = v.parse::<u64>().map_err(|_| InvalidValue {
+                    type_ref: type_model.into(),
+                    field: value_name.to_owned(),
+                })?;
+                let system_time = SystemTime::UNIX_EPOCH + Duration::from_millis(num);
+                Ok(ElementValue::Date(system_time))
+            }
+            (ValueType::Boolean, JsonElement::String(v)) => match v.as_str() {
+                "0" => Ok(ElementValue::Bool(false)),
+                "1" => Ok(ElementValue::Bool(true)),
+                _ => invalid_value(),
+            },
+            (ValueType::GeneratedId, JsonElement::String(v)) => Ok(ElementValue::GeneratedId(v)),
+            (ValueType::CustomId, JsonElement::String(v)) => Ok(ElementValue::CustomId(v)),
+            (ValueType::CompressedString, JsonElement::String(_)) => {
+                unimplemented!("compressed string")
```

Comment:
not sure, we probably won't need it for a while

Reply:
we rarely if ever need IDs in binary form. In most cases we just pass them on as-is. I think it might be easier for language bindings for now but we can maybe put it up to refinement.

Reply:
I think "how do we represent IDs on the client" is a valid question for refinement